File: frontend/src/services/bookService.js
import api from './api';

const bookService = {
  getAllBooks: () => api.get('/books'),

  getBookById: (id) => api.get(`/books/${id}`),

  searchByTitle: (title) => api.get('/books/search/title', { params: { title } }),

  searchByAuthor: (author) => api.get('/books/search/author', { params: { author } }),

  searchByCategory: (category) => api.get('/books/search/category', { params: { category } }),

  getAvailableBooks: () => api.get('/books/available'),

  createBook: (book) => api.post('/books', book),

  updateBook: (id, book) => api.put(`/books/${id}`, book),

  deleteBook: (id) => api.delete(`/books/${id}`),
};

export default bookService;

File: backend/src/main/java/com/library/service/BookService.java
package com.library.service;

import com.library.dto.BookDTO;
import com.library.model.Book;

import java.util.List;

/**
 * Інтерфейс для роботи з книгами (Interface Segregation Principle)
 */
public interface BookService {

    BookDTO createBook(BookDTO bookDTO);

    BookDTO getBookById(Long id);

    BookDTO getBookByIsbn(String isbn);

    List<BookDTO> getAllBooks();

    List<BookDTO> searchBooksByTitle(String title);

    List<BookDTO> searchBooksByAuthor(String authorName);

    List<BookDTO> searchBooksByCategory(String categoryName);

    List<BookDTO> getAvailableBooks();

    BookDTO updateBook(Long id, BookDTO bookDTO);

    void deleteBook(Long id);

    void updateBookAvailability(Long bookId, int change);
}
